### DISCUSSION TITLE
Add an equivalent to useDebouncedValue for uncontrolled elements

### DISCUSSION THREAD

**Discussion by nmay231** [mantinedev/mantine#1679] [Feature requests]
**Feature Request**

While `useDebouncedValue` is good for many use cases, it only works for controlled inputs, requiring many rerenders. I suggest adding a variant for uncontrolled inputs (tentatively called `useDebouncedState`).

[Here is a repo](https://github.com/nmay231/use-debounced-state-demo) with a demo implementation that compares the number of renders side by side. The important files are [`src/App.tsx`](https://github.com/nmay231/use-debounced-state-demo/blob/master/src/App.tsx) and [`src/useDebouncedState.ts`](https://github.com/nmay231/use-debounced-state-demo/blob/master/src/useDebouncedState.ts).

**Example**

Psuedo-code showing the different APIs:

`useDebouncedValue` (existing hook):

```typescript
const [_state, setState] = useState("default");
// Causes a render on every keypress
const [state] = useDebouncedValue(_state, 500);

return (
    <>
        <input value={_state} onChange={(e) => setState(e.target.value)} />
        <p>Value: {state}</p>
    </>
);
```

`useDebouncedState` (proposed additional hook):
```typescript
// Only causes a render when the value "debounces"
const [state, setState] = useDebouncedState("default", 500);

return (
    <>
        <input defaultValue={state} onChange={(e) => setState(e.target.value)} />
        <p>Value: {state}</p>
    </>
);
```

**Note**

If you think about it, `useDebouncedState` cannot have the exact same signature as `useDebouncedValue` because it must provide its own setter function. That's why the return signature is based on `useState` (and why I gave it its name).

**Comment by nmay231** [comment]
Here's a code-sandbox of it: https://codesandbox.io/s/keen-meadow-4kqqlv.

An alternative or additional API could be

```javascript
const [state, setState] = useDebounced(useState("default"), 500)
```

This allows it to work with external state libraries like [jotai](https://github.com/pmndrs/jotai)'s `useAtom`.

**Comment by rtivital** [comment]
Seems fine, feel free to open a PR with new hook

**Reply by nmay231** [reply]
Sounds good. Which API would you prefer? I prefer 1 or 3.

1) Require a state provider like `useState` or `useAtom`.
```javascript
const [value, setValue] = useDebounced(useState("default"), {...config})
```

2) Don't allow a custom state provider, i.e. always use `useState`.
```javascript
const [value, setValue] = useDebouncedState("default", {...config})
```

3) Allow both. This means we have to detect if they passed in a state provider, and the easiest way I can think of is to disallow array states (arrays would have to be nested inside an object, or be passed using `useState`).
```javascript
const [value, setValue] = useDebouncedState({ array: [1, 2, 3] }, {...config})
// or
const [value, setValue] = useDebouncedState(useState([1, 2, 3]), {...config})
```

**Reply by rtivital** [reply]
second option seems the best
